Automattic-Tracks-iOS/Services/EventLoggingUploadQueue.swift

```diff
@@ -0,0 +1,43 @@
+import Foundation
+
+open class EventLoggingUploadQueue {
+
+    private let fileManager = FileManager.default
+
+    /// The log file on top of the queue
+    var first: LogFile? {
+
+        guard let url = try? fileManager.contentsOfDirectory(at: storageDirectory).first else {
+            return nil
+        }
+
+        return LogFile(url: url, uuid: url.lastPathComponent)
```

Comment:
I am sure this would be fine for a lot of people who'd review this PR, but the thing is, this little bit of code assumes knowledge of how something is implemented in another part of the library. In these type of cases, I always suggest refactoring the code a bit to remove the extra cognitive load.

In this case, the storage path creation happens in this class, so I'd be totally fine to leave it as is. However, if you think it'd be a good improvement, I'd suggest maybe an extension method to `LogFile` in this file with 2 methods, one creating the `storageUrl` and the other one creating a logfile from a given storage url, so the pairing is in the same place.

Again, I am just sharing how I think about these things and a possible improvement that might make an impact in a lot of cases but has a very limited one in this case. Hope that makes sense!

Reply:
Yeah this was a tricky one! I'm not big on classes having any input into their persistence, so this seemed clean. My thought was that if we needed to persist it elsewhere we could have a `LogFileStore` that wrangles that, but it'd be overkill to have it for just one instance. 

I've reworked it to allow initialization without this implementation detail leaking, as well as adding a corresponding `fileName` property that consumers can use to persist it. This hides the implementation details a bit more nicely. 

WDYT? Implemented in 0f92472a683191817f3847479b0959fdc09e2cbd.

Reply:
I think with this new implementation makes the following problem mentioned in the original comment more apparent:

> this little bit of code assumes knowledge of how something is implemented in another part of the library

Having said that, I think the only issue is the placement of the new method. So, how about we have a (private?) class with 2 methods for creating the `LogFile` and `storageURL`, is that maybe what you had in mind with `LogFileStore`?

```
private final class LogFileStore {
    func createLogFileFromURL(_ url: URL) -> LogFile {}
    func createStorageURLForLogFile(in: StorageDirectory, _ logFile: LogFile) -> String
}
```
_Wrote this on Github and although I can read Swift, that doesn't mean I can write it, so please imagine this more Swifty and with correct syntax 🙈_

It doesn't have to be a class, although I think there is some benefit to making it a class/protocol, the most important thing I was trying to convey here is to make these the bits of logic live together (possibly encapsulated) so we don't assume one in the other.

---

Now I am doubting if this discussion was really necessary 😬 @shiki Any thoughts on this discussion, am I maybe thinking way too hard about this?

Reply:
The way I understood this was that `EventLoggingUploadQueue` manages `LogFiles` that are inside a single directory. However, since a `LogFile` can be created by consumers, **there is no guarantee** that the `LogFile` passed to `remove()` is managed by `EventLoggingUploadQueue` itself. We currently handle that silently. 

I think you're right on the money here, @oguzkocer. I'd propose to even hide `Log.fileName` inside this `EventLoggingUploadQueue` as it is one of the queueing + storage implementation detail. 

But I find it challenging to give recommendations right now until I see how everything fits together — which I hope to see in upcoming PRs. Perhaps we could do an architectural review once all the classes have been put together?

Reply:
> Perhaps we could do an architectural review once all the classes have been put together?

That sounds good to me 👍